cmd/zdb/zdb.c

```diff
@@ -6578,14 +6618,28 @@ zdb_read_block(char *thing, spa_t *spa)
 		borrowed = B_TRUE;
 	}
 
+	/* Try to detect invalid block pointer. */
+	if ((flags & ZDB_FLAG_PRINT_BLKPTR || flags & ZDB_FLAG_INDIRECT) &&
+	    !(flags & ZDB_FLAG_DECOMPRESS)) {
```

Comment:
The warning is helpful, but perhaps we can do better.  We have this nice `zdb_blkptr_verify()` function to sanity check block pointers.  I'd argue that for `-i` and `-b` it would make sense to automatically retry with `-d` when `zdb_blkptr_verify()` fails and only then print an error when both fail.  That's one less thing to have to remember when debugging.

Reply:
Good idea on retry with -d.  I'll also see how hard it is to refactor zfs_blkptr_verify to avoid the code duplication and let you know about that.

Reply:
I have made the suggested changes including refactoring zfs_blkptr_verify.